numpyro/distributions/continuous.py

```diff
@@ -1188,6 +1192,141 @@ def infer_shapes(
         return batch_shape, event_shape
 
 
+class CAR(Distribution):
+    r"""
+    The Conditional Autoregressive (CAR) distribution is a special case of the multivariate
+    normal in which the precision matrix is structured according to the adjacency matrix of
+    sites. The amount of autocorrelation between sites is controlled by :math:`\alpha`. The
+    distribution is a popular prior for areal spatial data.
+
+    :param float or ndarray loc: mean of the multivariate normal
+    :param float alpha: autoregression parameter. For most cases, the value should lie
+        between 0 (sites are independent, collapses to an iid multivariate normal) and
+        1 (perfect autocorrelation between sites), but the specification allows for negative
+        correlations.
+    :param float tau: positive precision for the multivariate normal
+    :param ndarray or scipy.sparse.csr_matrix W: symmetric adjacency matrix where 1
+        indicates adjacency between sites and 0 otherwise
+    :param bool is_sparse: whether to use a sparse form of W in calculations (must be True is W is a
+        :class:`scipy.sparse.spmatrix`)
+    """
+
+    arg_constraints = {
+        "loc": constraints.real_vector,
+        "alpha": constraints.open_interval(-1, 1),
+        "tau": constraints.positive,
+        "W": constraints.dependent_property(is_discrete=False, event_dim=2),
+    }
+    support = constraints.real_vector
+    reparametrized_params = [
+        "loc",
+        "alpha",
+        "tau",
+        "W",
+    ]
+
+    def __init__(self, loc, alpha, tau, W, *, is_sparse=False, validate_args=None):
+        if jnp.ndim(loc) == 0:
+            (loc,) = promote_shapes(loc, shape=(1,))
+
+        self.is_sparse = is_sparse
+
+        if self.is_sparse:
+            if not sparse.issparse(W):
+                assert isinstance(W, np.ndarray)
+            self.W = sparse.csr_matrix(W)
+        else:
+            assert not sparse.issparse(W)
+            self.W = W
+
+        batch_shape = lax.broadcast_shapes(
+            jnp.shape(loc)[:-1],
+            jnp.shape(alpha),
+            jnp.shape(tau),
+            jnp.shape(self.W)[:-2],
+        )
+        event_shape = jnp.shape(self.W)[-1:]
+        (self.loc,) = promote_shapes(loc, shape=batch_shape + event_shape)
+        self.alpha, self.tau = promote_shapes(alpha, tau, shape=batch_shape)
+
+        super(CAR, self).__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+        if self._validate_args:
+            assert (
+                self.W.sum(axis=-1) > 0
+            ), "all sites in adjacency matrix W must have neighbours"
+
+            if sparse.issparse(self.W):
+                assert (self.W != self.W.T).nnz == 0, "W must be symmetric"
+            else:
```

Comment:
Could you change this to only check if self.W is a numpy array? I don't think the code works for jax arrays/abstract arrays/tracers.

Reply:
`np.array_equal(jnp.asarray(W), jnp.asarray(W).T)` works fine so I will leave it as is

Reply:
Could you try to `jax.jit` any function using CAR with validation? I think some errors will happen. You can perform numpy operators on jax arrays outside of jit context but cannot inside a jit context. In principle, we should only use jax operators on jax arrays. You can try to jit something like
```
import jax
import numpy

@jax.jit
def f():
  x = jax.numpy.ones(10)
  assert numpy.all(x)
  return x

f()
```

Reply:
How about just using `(self.W == self.W.T).all()` instead of `np.array_equal()`. That will be jnp/np agnostic

Reply:
Even so, `assert jax_array` will raise errors under jitting. A way to think about this is to consider jax arrays as abstract arrays which do not store any concrete values. Under jit, we only know a few things of a jax array: shape, dtype.